---
title: 'Germline Analysis Of NA12878 With Sarek'
date: 2020-05-08
permalink: /posts/2020/05/NA12878_germline/
tags:
  - NA12878
  - sarek
  - germline
---

NA12878 is likely the most sequenced whole genome sample, used for many benchmarks. We know quite a lot about SNVs, indels and structural variants to be found in this sample. This time I run a benchmark study using [Sarek](http://nf-co.re/sarek/) a pipeline for somatic and germline analysis written in [nextflow](http://nextflow.io).

### NA12878, GiaB and Sarek

The goal of [GiaB](https://www.nist.gov/programs-projects/genome-bottle) a.k.a the Genome in a Bottle project is to provide an "authoritative characterization of human genomes for use in analytical validation and technology development". By providing validated polymorphisms and short reads this is the first stop for practicing bioinformaticians to validate software used for genomics. Our pipeline Sarek is a good example for such a validation: we did not develop our own aligner and/or variant caller, etc. to process normal/tumor samples, but made a collection on software and best practices (based on [GATK](https://gatk.broadinstitute.org/hc/en-us/sections/360007226651-Best-Practices-Workflows)) and linked the parts together to have a complete pipeline from raw FASTQ to annotated variant calls. 

In this post I will review the germline-only processing step: how to get raw data, how to run Sarek with GRCh38, where to find germline results and how to compare the two germline callers ([HaplotypeCaller](https://gatk.broadinstitute.org/hc/en-us/articles/360042913231-HaplotypeCaller) and [Strelka](https://github.com/Illumina/strelka)) to the reference call set regarding SNVs and small indels.

### Getting data

Links to GiaB raw Illumina reads (and many other files) are on github at the [genome-in-a-bottle](https://github.com/genome-in-a-bottle/giab_data_indexes) project pages. I have downloaded only some parts, not the whole 300x coverage data: 

```
for d in 131219_D00360_005_BH814YADXX 131219_D00360_006_AH81VLADXX; do
  wget --mirror --no-parent ftp://ftp-trace.ncbi.nlm.nih.gov/giab/ftp/data/NA12878/NIST_NA12878_HG001_HiSeq_300x/$d/
done
```

That resulted in 129G raw fastq file. To get a TSV file as an input for Sarek we have to have the read groups and the gender right (actually gender is not used in germline analysis). Nevertheless, if our files are at the `/NA12878/fastq` directory, we can make a TSV input definition file for Sarek as: 

```
find /NA12878/fastq -name *R1*fastq.gz| sort| \
awk '{ \
  split($1,f,"/"); \
  split(f[11],c,"_"); \
  split(f[14],bn,"."); \
  r2=$1; \
  sub("_R1_","_R2_",r2); \
  printf("NA12878\tXX\t0\tNA12878N\t%s_%s\t%s\t%s\n",c[4],bn[1],$1,r2); \
}' > na12878.tsv
```

### Processing with Sarek

I am getting these line from files generated by Sarek itself and saved in the `results/pipeline_info` directory. The `pipeline_report.txt` file shows the workflow for germline only have been started with parameters below. Instead of `-profile munin` one can use the `singularity` or the `docker` profile, custom profiles can be stored locally or to be stored at https://github.com/nf-core/configs . The command line to start the analysis:

```
$ nextflow run nf-core/sarek -r 2.6 -profile munin --steps mapping,variantcalling --tools haplotypecaller,strelka --input na12878.tsv
```

Running all the mapping and variant calling took 1d 1h 49m on this sample (coverage is about 46X) with 48 CPUs. Results are ins the `results` directory; BAMs in the `Preprocessing`, VCFs containing mutation calls in the `VariantCalling` directory:

```
results/
├── pipeline_info
├── Preprocessing
│   ├── NA12878N
│   │   ├── DuplicatesMarked
│   │   └── Recalibrated
│   └── TSV
│
└── VariantCalling
    └── NA12878N
        ├── HaplotypeCaller
        ├── HaplotypeCallerGVCF
        └── Strelka
```

Besides the pipeline info there is a `Reports` directory (now omitted) containing quality control data. 

### Comparing Variant Callers

```
~/dev/btb-scripts/gatkVQSR.sh HaplotypeCaller_NA12878N.vcf.gz
zcat HaplotypeCaller_NA12878N.SNP.INDEL.recalibrated.vcf.gz| awk '/^#/{print}!/^#/&&/PASS/{print}' | bgzip -@ 12 -c > HaplotypeCaller_NA12878N.SNP.INDEL.PASS.recalibrated.vcf.gz
tabix HaplotypeCaller_NA12878N.SNP.INDEL.PASS.recalibrated.vcf.gz
$ zcat HaplotypeCaller_NA12878N.SNP.INDEL.PASS.recalibrated.vcf.gz| grep -c PASS 
4218255
zcat Strelka_NA12878N_variants.vcf.gz| awk '/^#/{print}!/^#/&&/PASS/{print}' | bgzip -@ 12 -c > Strelka_NA12878N_variants.PASS.vcf.gz
tabix Strelka_NA12878N_variants.PASS.vcf.gz
$ zcat Strelka_NA12878N_variants.PASS.vcf.gz| grep -c PASS
4635536

$ vcftools --gzvcf Strelka_NA12878N_variants.PASS.vcf.gz --bed HG001_GRCh38_GIAB_highconf_CG-IllFB-IllGATKHC-Ion-10X-SOLID_CHROM1-X_v.3.3.2_highconf_nosomaticdel_noCENorHET7.bed --out Strelka.GiaB --recode --recode-INFO-all
[... some format warnings ...]
Outputting VCF file...
        Read 438100 BED file entries.
After filtering, kept 3540712 out of a possible 4635535 Sites
Run Time = 539.00 seconds

$ vcftools --vcf HaplotypeCaller_NA12878N.SNP.INDEL.PASS.recalibrated.vcf --bed HG001_GRCh38_GIAB_highconf_CG-IllFB-IllGATKHC-Ion-10X-SOLID_CHROM1-X_v.3.3.2_highconf_nosomaticdel_noCENorHET7.bed --out HaplotypeCaller.GiaB --recode --recode-INFO-all

Outputting VCF file...
        Read 438100 BED file entries.
After filtering, kept 3353222 out of a possible 4218254 Sites
Run Time = 487.00 seconds


$ mkdir snps
$ awk '/^#/{print}!/^#/&&length($4)==length($5){print}' Strelka.GiaB.recode.vcf > snps/Strelka.SNPs.vcf
$ awk '/^#/{print}!/^#/&&length($4)==length($5){print}' HaplotypeCaller.GiaB.recode.vcf > snps/HaplotypeCaller.SNPS.vcf
$ zcat HG001_GRCh38_GIAB_highconf_CG-IllFB-IllGATKHC-Ion-10X-SOLID_CHROM1-X_v.3.3.2_highconf_PGandRTGphasetransfer.vcf.gz |awk '/^#/{print}!/^#/&&length($4)==length($5){print}' > snps/GiaB.SNPS.vcf
$ cd snps
snps $ for f in *vcf; do echo -n $f" ";awk '!/^#/{print}' $f| wc -l ; done 
GiaB.SNPS.vcf 3088156
HaplotypeCaller.SNPS.vcf 2867483
Strelka.SNPs.vcf 3046192

$ for f in *vcf; do bgzip -@32 $f; tabix ${f}.gz; done
$ bcftools isec GiaB.SNPS.vcf.gz HaplotypeCaller.SNPS.vcf.gz -p Giab.HC

Giab.HC/0000.vcf        for records private to  GiaB.SNPS.vcf.gz
Giab.HC/0001.vcf        for records private to  HaplotypeCaller.SNPS.vcf.gz
Giab.HC/0002.vcf        for records from GiaB.SNPS.vcf.gz shared by both
Giab.HC/0003.vcf        for records from HaplotypeCaller.SNPS.vcf.gz shared by both

snps/Giab.HC $ for f in *vcf; do echo -n $f" "; awk '!/^#/{print}' $f|wc -l ; done
0000.vcf 226886       <- not found by HC 
0001.vcf 6213         <- not in GiaB
0002.vcf 2861270      <-
0003.vcf 2861270      <- common SNPs
$ cd ..
$ bcftools isec GiaB.SNPS.vcf.gz Strelka.SNPs.vcf.gz -p Giab.Strelka
$ snps $ cd Giab.Strelka/
$ snps/Giab.Strelka $ for f in *vcf; do echo -n $f" "; awk '!/^#/{print}' $f|wc -l ; done
0000.vcf 47533
0001.vcf 5569
0002.vcf 3040623
0003.vcf 3040623
$ cd ..
snps $ bcftools isec HaplotypeCaller.SNPS.vcf.gz Strelka.SNPs.vcf.gz -p HC.Strelka
snps $ cd HC.Strelka/
snps/HC.Strelka $ for f in *vcf; do echo -n $f" "; awk '!/^#/{print}' $f|wc -l ; done
0000.vcf 6674          <- HC only
0001.vcf 185383        <- Strelka only
0002.vcf 2860809       <- common
0003.vcf 2860809

$ bgzip -c Giab.Strelka/0003.vcf > GS.common.gz && tabix GS.common.gz
$ bcftools isec GS.common.gz HaplotypeCaller.SNPS.vcf.gz -p GS.common.HC 
$ bgzip -c Giab.HC/0003.vcf > GH.common.gz && tabix GH.common.gz
$ bcftools isec GH.common.gz Strelka.SNPS.vcf.gz -p GH.common.S 
$ bgzip -c HC.Strelka/0003.vcf > HS.common.gz && tabix HS.common.gz
$ bcftools isec HS.common.gz GiaB.SNPS.vcf.gz -p HS.common.G 

snps $ for f in GS.common.HC/*vcf; do echo -n $f" "; awk '!/^#/{print}' $f|wc -l ; done
GS.common.HC/0000.vcf 180954  <- common for GiaB and Strelka
GS.common.HC/0001.vcf 7814    <- only in HaplotypeCaller
GS.common.HC/0002.vcf 2859669 <- common for all
GS.common.HC/0003.vcf 2859669
snps $ for f in GH.common.S/*vcf; do echo -n $f" "; awk '!/^#/{print}' $f|wc -l ; done
GH.common.S/0000.vcf 1601    <- common for GiaB and HaplotypeCaller
GH.common.S/0001.vcf 186523  <- only in Strelka
GH.common.S/0002.vcf 2859669 <- common for all
GH.common.S/0003.vcf 2859669
snps $ for f in HS.common.G/*vcf; do echo -n $f" "; awk '!/^#/{print}' $f|wc -l ; done
HS.common.G/0000.vcf 1140    <- common for HaplotypeCaller and Strelka 
HS.common.G/0001.vcf 228487  <- only in GiaB
HS.common.G/0002.vcf 2859669 <- common for all
HS.common.G/0003.vcf 2859669

```
<br/><img src='/images/GiaB.germline.png'>
