---
title: 'Germline Analysis Of NA12878 With Sarek'
date: 2020-05-08
permalink: /posts/2020/05/NA12878_germline/
tags:
  - NA12878
  - sarek
  - germline
---

NA12878 is likely the most sequenced whole genome sample, used for many benchmarks. We know quite a lot about SNVs, indels and structural variants to be found in this sample. This time I run a benchmark study using [Sarek](http://nf-co.re/sarek/) a pipeline for somatic and germline analysis written in [nextflow](http://nextflow.io).

NA12878, GiaB and Sarek
======

The goal of [GiaB](https://www.nist.gov/programs-projects/genome-bottle) a.k.a the Genome in a Bottle project is to provide an "authoritative characterization of human genomes for use in analytical validation and technology development". By providing validated polymorphisms and short reads this is the first stop for practicing bioinformaticians to validate software used for genomics. Our pipeline Sarek is a good example for such a validation: we did not develop our own aligner and/or variant caller, etc. to process normal/tumor samples, but made a collection on software and best practices (based on [GATK](https://gatk.broadinstitute.org/hc/en-us/sections/360007226651-Best-Practices-Workflows)) and linked the parts together to have a complete pipeline from raw FASTQ to annotated variant calls. 

In this post I will review the germline-only processing step: how to get raw data, how to run Sarek with GRCh38, where to find germline results and how to compare the two germline callers ([HaplotypeCaller](https://gatk.broadinstitute.org/hc/en-us/articles/360042913231-HaplotypeCaller) and [Strelka](https://github.com/Illumina/strelka)) to the reference call set regarding SNVs and small indels.

Getting data
======

Links to GiaB raw Illumina reads (and many other files) are on github at the [genome-in-a-bottle](https://github.com/genome-in-a-bottle/giab_data_indexes) project pages. I have downloaded only some parts, not the whole 300x coverage data: 

```
for d in 131219_D00360_005_BH814YADXX 131219_D00360_006_AH81VLADXX; do
  wget --mirror --no-parent ftp://ftp-trace.ncbi.nlm.nih.gov/giab/ftp/data/NA12878/NIST_NA12878_HG001_HiSeq_300x/$d/
done
```

That resulted in 129G raw fastq file. To get a TSV file as an input for Sarek we have to have the read groups and the gender right (actually gender is not used in germline analysis). Nevertheless, if our files are at the `/NA12878/fastq` directory, we can make a TSV input definition file for Sarek as: 

```
find /NA12878/fastq -name *R1*fastq.gz| sort| \
awk '{ \
  split($1,f,"/"); \
  split(f[11],c,"_"); \
  split(f[14],bn,"."); \
  r2=$1; \
  sub("_R1_","_R2_",r2); \
  printf("NA12878\tXX\t0\tNA12878N\t%s_%s\t%s\t%s\n",c[4],bn[1],$1,r2); \
}' > na12878.tsv
```

Processing with Sarek
=====


Comparing Variant Callers
=====
