---
title: 'Visualizing Control-FREEC BAF and LogR data with IGV'
date: 2020-05-07
permalink: /posts/2020/05/control-freec-igv/
tags:
  - CNV
  - Control-FREEC
  - structural variants
---
## Visualizing Control-FREEC BAF and LogR data with IGV

[Control-FREEC](http://boevalab.inf.ethz.ch/FREEC/) is a software "for detection of copy-number changes and allelic imbalances (including LOH) using deep-sequencing data" according to their homepage. Although this page is about visualizing Control-FREEC data, a similar software called [ASCAT](https://www.crick.ac.uk/research/labs/peter-van-loo/software) is also used in Sarek for CNV/LOH/ploidy and cell purity estimation. Explanation of BAF and LogR can be found at doi:10.1007/978-1-61779-400-1_4 .

[Integrative Genomics Viewer a.k.a. IGV](https://igv.org/) is not the best, not the fastest, but likely the most popular browser for NGS data. Good news is that besides BAM, BED and VCF files it can also show BedGraph format files. 

Control-FREEC results are saved by Sarek in the `results/VariantCalling/TUMOR_vs_NORMAL/Control-FREEC` directory, and there are two .BedGraph entries: `*.pileup_ratio.BedGraph` contains LogR values for the tumor sample `*.pileup_normal_ratio.BedGraph` is the same, well, for the normal sample. 

The [BedGraph](https://genome.ucsc.edu/goldenPath/help/bedgraph.html) format itself is an extension of BED: have to begin with a track description line and values are listed as `chromosome start end value` delimited by tabs. This way we can visualize LogR values:

```
track type=bedGraph name="copy neutral LOH from T.pileup.gz_ratio.BedGraph" description="Copy Neutral LOH" color=100,100,100
chr1    124950000       125000000       3.5375
chr1    125000000       125050000       3.43962
chr1    150750000       150800000       3.0908
chr1    150800000       150850000       3.10692
chr1    150850000       150900000       2.69283
chr1    150900000       150950000       3.04243
chr1    150950000       151000000       3.19796
```

It is straightforward to load these two files to IGV, but first we will be disappointed, as we will see (if at all) only two thin bright green lines (here shown for chr3 of HeLa cells):

<br/><img src='/images/HeLa_chr3_logr_default.png'>

So, _right_-click on the track, and 
 - Change track height to 200 
 - Change track color to blue or red (only for positive values)
 - use Points instead of bars
 - use Autoscale
 
and do it for both tracks. In the picture below I use blue for the normal, and red for the tumour tracks:

<br/><img src='/images/HeLa_blue_red_logr.png'>

Now we can zoom in and have a better look at the ratio changes.


Processing BAF data to be read by IGV
------
to be continued ...
