---
title: 'Germline Analysis Of NA12878 With Sarek'
date: 2020-05-08
permalink: /posts/2020/05/NA12878_germline/
tags:
  - NA12878
  - sarek
  - germline
---

NA12878 is likely the most sequenced whole genome sample, used for many benchmarks. We know quite a lot about SNVs, indels and structural variants to be found in this sample. This time I run a benchmark study using [Sarek](http://nf-co.re/sarek/) a pipeline for somatic and germline analysis written in [nextflow](http://nextflow.io).

NA12878, GiaB and Sarek
======

The goal of [GiaB](https://www.nist.gov/programs-projects/genome-bottle) a.k.a the Genome in a Bottle project is to provide an "authoritative characterization of human genomes for use in analytical validation and technology development". With providing validated polymorphisms and short reads this is the first stop for practicing bioinformaticians to validate software used for genomics. Our pipeline Sarek is a good example for such a validation: we did not develop our own aligner and/or variant caller, etc. to process normal/tumor samples, but made a collection on software and best practices (based on [GATK](https://gatk.broadinstitute.org/hc/en-us/sections/360007226651-Best-Practices-Workflows)) and linked the parts together to have a complete pipeline from raw FASTQ to annotated variant calls. 

In this post I will review the germline-only processing step: how to get raw data, how to run Sarek with GRCh38, where to find germline results and how to compare the two germline callers ([HaplotypeCaller](https://gatk.broadinstitute.org/hc/en-us/articles/360042913231-HaplotypeCaller) and [Strelka](https://github.com/Illumina/strelka)) to the reference call set regarding SNVs and small indels.

Getting data
======


Processing with Sarek
=====


Comparing Variant Callers
=====
