---
title: 'Germline analysis of NA12878 with Sarek'
date: 2020-05-08
permalink: /posts/2020/05/blog-post-5/
tags:
  - NA12878
  - sarek
  - germline
---

NA12878 is likely the most sequenced whole genome sample.

NA12878 and GiaB
======

Getting data and processing with Sarek
======

Singularity:
```
$ docker login
Login with your Docker ID to push and pull images from Docker Hub. If you don't have a Docker ID, head over to https://hub.docker.com to create one.
Username (szilvajuhos):
Password:
Login Succeeded
$ singularity pull docker://shollizeck/bamsurgeon
WARNING: Authentication token file not found : Only pulls of public images will succeed
INFO:    Starting build...
Getting image source signatures
Copying blob sha256:3386e6af03b043219225367632569465e5ecd47391d1f99a6d265e51bd463a83
 42.08 MiB / 42.08 MiB [====================================================] 0s
[...]

$ singularity shell bamsurgeon_latest.sif
Singularity bamsurgeon_latest.sif:/data2/NA12878/bamsurgeon> addsnv.py
usage: addsnv.py [-h] -v VARFILENAME -f BAMFILENAME -r REFFASTA -o OUTBAMFILE
                 [-s SNVFRAC] [-m MUTFRAC] [-n NUMSNVS] [-c CNVFILE]
                 [-d COVERDIFF] [-z HAPLOSIZE] [-p PROCS]
                 [--picardjar PICARDJAR] [--mindepth MINDEPTH]
                 [--maxdepth MAXDEPTH] [--minmutreads MINMUTREADS]
                 [--avoidreads AVOIDREADS] [--nomut] [--ignoresnps]
                 [--ignoreref] [--force] [--insane] [--single]
                 [--maxopen MAXOPEN] [--requirepaired] [--tagreads]
                 [--skipmerge] [--ignorepileup] [--aligner ALIGNER]
                 [--alignopts ALIGNOPTS] [--tmpdir TMPDIR] [--seed SEED]
addsnv.py: error: argument -v/--varfile is required

```
So far so good.

```
$ find /data1/P2233 -type f -name "Strelka\*snvs\*ann\*vcf"| grep CADD | xargs grep HIGH | tee raw.snps| wc -l 
1393
$ awk -F: 'BEGIN{OFS="\t"}$2!~/#/{split($2,cols,"\t");print cols[1],cols[2],cols[2],0.5}' raw.snps| sort -u -V -k1,1 -k2,2n| tee SNPS.tsv |wc -l
671
$ samtools view -@36 -b -o NA12878.chr17.bam ../results/Preprocessing/NA12878N/Recalibrated/NA12878N.recal.bam chr17:
$ samtools index NA12878.chr17.bam
$ mkdir ref &&  cd ref
$ for f in /references/igenomes/Homo\_sapiens/GATK/GRCh38/Sequence/BWAIndex/\* ; do ln -s $f ; done
$ for f in /references/igenomes/Homo\_sapiens/GATK/GRCh38/Sequence/WholeGenomeFasta/\* ; do ln -s $f ; done
$ cd ..
$ singularity shell bamsurgeon\_latest.sif
Singularity bamsurgeon\_latest.sif:/data2/NA12878/bamsurgeon> time addsnv.py -v chr17.snps.tsv -f NA12878.chr17.bam -r ref/Homo\_sapiens\_assembly38.fasta -o NA12878.chr17.somatic.bam -p 12 --force
[...]
real    10m54.014s
user    19m3.553s
sys     6m49.376s

Singularity bamsurgeon\_latest.sif:/data2/NA12878/bamsurgeon> time addsnv.py -v SNPS.tsv -f ../results/Preprocessing/NA12878N/Recalibrated/NA12878N.recal.bam -r ref/Homo\_sapiens\_assembly38.fasta -o NA12878.somatic.bam -p 12 --force
```



Comparing Variant Callers
-----
