---
title: 'Germline Analysis Of NA12878 With Sarek'
date: 2020-05-18
permalink: /posts/2020/05/NA12878_germline/
tags:
  - NA12878
  - sarek
  - germline
---

NA12878 is likely the most sequenced whole genome sample, used for many benchmarks. We know quite a lot about SNVs, indels and structural variants to be found in this sample. This time I run a benchmark study using [Sarek](http://nf-co.re/sarek/) a pipeline for somatic and germline analysis written in [nextflow](http://nextflow.io).

### NA12878, GiaB and Sarek

The goal of [GiaB](https://www.nist.gov/programs-projects/genome-bottle) a.k.a the Genome in a Bottle project is to provide an "authoritative characterization of human genomes for use in analytical validation and technology development". By providing validated polymorphisms and short reads this is the first stop for practicing bioinformaticians to validate software used for genomics. Our pipeline Sarek is a good example for such a validation: we did not develop our own aligner and/or variant caller, etc. to process normal/tumor samples, but made a collection on software and best practices (based on [GATK](https://gatk.broadinstitute.org/hc/en-us/sections/360007226651-Best-Practices-Workflows)) and linked the parts together to have a complete pipeline from raw FASTQ to annotated variant calls. 

In this post I will review the germline-only processing step: how to get raw data, how to run Sarek with GRCh38, where to find germline results and how to compare the two germline callers ([HaplotypeCaller](https://gatk.broadinstitute.org/hc/en-us/articles/360042913231-HaplotypeCaller) and [Strelka](https://github.com/Illumina/strelka)) to the reference call set regarding SNVs and small indels.

### Getting data

Links to GiaB raw Illumina reads (and many other files) are on github at the [genome-in-a-bottle](https://github.com/genome-in-a-bottle/giab_data_indexes) project pages. I have downloaded only some parts, not the whole 300x coverage data: 

```
for d in 131219_D00360_005_BH814YADXX 131219_D00360_006_AH81VLADXX; do
  wget --mirror --no-parent ftp://ftp-trace.ncbi.nlm.nih.gov/giab/ftp/data/NA12878/NIST_NA12878_HG001_HiSeq_300x/$d/
done
```

That resulted in 129G raw fastq file. If you are looking for accession numbers from ENA, the best is to search for the library name (like 131219_D00360_005_BH814YADXX). Accession numbers for 131219_D00360_006_AH81VLADXX are SRX1049774 -SRX1049779, for 131219_D00360_005_BH814YADXX SRR2052337 are SRR2052339, SRR2052342, SRR2052345, SRR20523428. To get a TSV file as an input for Sarek we have to have the read groups and the gender right (actually gender is not used in germline analysis). Nevertheless, if our files are at the `/NA12878/fastq` directory, we can make a TSV input definition file for Sarek as: 

```
find /NA12878/fastq -name *R1*fastq.gz| sort| \
awk '{ \
  split($1,f,"/"); \
  split(f[11],c,"_"); \
  split(f[14],bn,"."); \
  r2=$1; \
  sub("_R1_","_R2_",r2); \
  printf("NA12878\tXX\t0\tNA12878N\t%s_%s\t%s\t%s\n",c[4],bn[1],$1,r2); \
}' > na12878.tsv
```

### Processing with Sarek

I am getting these line from files generated by Sarek itself and saved in the `results/pipeline_info` directory. The `pipeline_report.txt` file shows the parameters used for starting the germline analysis. Instead of `-profile munin` one can use the `singularity` or the `docker` profile. Custom profiles can be stored locally or to be stored at https://github.com/nf-core/configs . The command line to start the analysis:

```
$ nextflow run nf-core/sarek -r 2.6 \
  -profile munin \
  --steps mapping,variantcalling \
  --tools haplotypecaller,strelka \
  --input na12878.tsv
```

Running all the mapping and variant calling took 1d 1h 49m on this sample (coverage is about 46X) with 48 CPUs. Results are ins the `results` directory; BAMs in the `Preprocessing`, VCFs containing mutation calls in the `VariantCalling` directory:

```
results/
├── pipeline_info
├── Preprocessing
│   ├── NA12878N
│   │   ├── DuplicatesMarked
│   │   └── Recalibrated
│   └── TSV
│
└── VariantCalling
    └── NA12878N
        ├── HaplotypeCaller
        ├── HaplotypeCallerGVCF
        └── Strelka
```

Besides the pipeline info, there is a `Reports` directory (now omitted) containing quality control data. 

### Comparing Variant Callers

To get a relatively fair comparison, first of all we want to have only calls that are in high-confidence regions. We can get them - again - from the [GiaB github page](https://github.com/genome-in-a-bottle/giab_latest_release) pointing towards https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/release/NA12878_HG001/latest/GRCh38/. I am using version deposited at 2017-02-24 13:36 . The VCF file contains high confidence calls, and BED file for these high-confidence regions also can be downloaded. We will select only those HaplotypeCaller and Strelka calls that are falling into these segments.

*If you are interested in results, you can skip the thechnical details below, scroll down to the Venn-diagrams.* 

##### Variant Quality Score recalibration and filtering variants with the PASS flag:

Strelka adds a PASS filter at the variant call step, but HaplotypeCaller calls have to be recalibrated. For recalibration I have used a [small script](https://github.com/szilvajuhos/btb-scripts/blob/master/gatkVQSR.sh) I wrote earlier - Sarek will do [VQSR](https://github.com/nf-core/sarek/issues/89) also eventually. 

```
$ gatkVQSR.sh HaplotypeCaller_NA12878N.vcf.gz
$ zcat HaplotypeCaller_NA12878N.SNP.INDEL.recalibrated.vcf.gz| \
  awk '/^#/{print}!/^#/&&/PASS/{print}' | \
  bgzip -@ 12 -c > HaplotypeCaller_NA12878N.SNP.INDEL.PASS.recalibrated.vcf.gz 
$ tabix HaplotypeCaller_NA12878N.SNP.INDEL.PASS.recalibrated.vcf.gz
$ zcat Strelka_NA12878N_variants.vcf.gz| \
  awk '/^#/{print}!/^#/&&/PASS/{print}' | \
  bgzip -@ 12 -c > Strelka_NA12878N_variants.PASS.vcf.gz
$ tabix Strelka_NA12878N_variants.PASS.vcf.gz
```
There are about 4 million PASS-ed calls in both the HaplotypeCaller and Strelka sets as expected: 
```
$ zcat HaplotypeCaller_NA12878N.SNP.INDEL.PASS.recalibrated.vcf.gz| grep -c PASS 
4218255
$ zcat Strelka_NA12878N_variants.PASS.vcf.gz| grep -c PASS
4635536
```
#### Selecting high-confidence regions:

Now we can select calls that are overlapping with the high-confidence regions with the help of the [vcftools](https://vcftools.github.io/index.html) software. Note, the HG001_... vcf file contains only high-confidence calls. 

```
$ vcftools --gzvcf Strelka_NA12878N_variants.PASS.vcf.gz \
  --bed HG001_GRCh38_GIAB_highconf_CG-IllFB-IllGATKHC-Ion-10X-SOLID_CHROM1-X_v.3.3.2_highconf_nosomaticdel_noCENorHET7.bed \
  --out Strelka.GiaB --recode --recode-INFO-all
[... some format warnings ...]
Outputting VCF file...
        Read 438100 BED file entries.
After filtering, kept 3540712 out of a possible 4635535 Sites
Run Time = 539.00 seconds
$ vcftools --vcf HaplotypeCaller_NA12878N.SNP.INDEL.PASS.recalibrated.vcf \
  --bed HG001_GRCh38_GIAB_highconf_CG-IllFB-IllGATKHC-Ion-10X-SOLID_CHROM1-X_v.3.3.2_highconf_nosomaticdel_noCENorHET7.bed \
  --out HaplotypeCaller.GiaB --recode --recode-INFO-all
[... some format warnings ...]
Outputting VCF file...
        Read 438100 BED file entries.
After filtering, kept 3353222 out of a possible 4218254 Sites
Run Time = 487.00 seconds
```

#### Filtering SNPs only

To assess SNPs only we have to select SNP calls from all sets and ignoring indels. First part of the awk expression below keeps the VCF header, the other part compares the length of the fourth and the fifth column, keeping only SNPs:

```
$ mkdir snps
$ awk '/^#/{print}!/^#/&&length($4)==length($5){print}' Strelka.GiaB.recode.vcf > snps/Strelka.SNPs.vcf
$ awk '/^#/{print}!/^#/&&length($4)==length($5){print}' HaplotypeCaller.GiaB.recode.vcf > snps/HaplotypeCaller.SNPS.vcf
$ zcat HG001_GRCh38_GIAB_highconf_CG-IllFB-IllGATKHC-Ion-10X-SOLID_CHROM1-X_v.3.3.2_highconf_PGandRTGphasetransfer.vcf.gz |\
  awk '/^#/{print}!/^#/&&length($4)==length($5){print}' > snps/GiaB.SNPS.vcf
```

Now we can have a look at the number of SNP calls; clearly there are more SNPs than either HaplotypeCaller or Strelka have found, therefore, we are not going to get 100% sensitivity by using only one variant caller. To get intersections and number of common calls I will use bcftools in this test. To get this, I have to compress VCFs by bgzip and index by tabix:  

```
$ cd snps
snps $ for f in *vcf; do echo -n $f" ";awk '!/^#/{print}' $f| wc -l ; done 
GiaB.SNPS.vcf 3088156
HaplotypeCaller.SNPS.vcf 2867483
Strelka.SNPs.vcf 3046192

$ for f in *vcf; do bgzip -@32 $f; tabix ${f}.gz; done
```

#### Generating intersections:

GiaB [recommends](https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/release/NA12878_HG001/latest/README_NISTv3.3.2.txt) to use [RTG tools](https://github.com/RealTimeGenomics/rtg-tools) for VCF comparison, but for my purposes [bcftools](https://samtools.github.io/bcftools/bcftools.html) will be just all right. When using bcftools, in the command line for intersection we have to supply a directory also where the results are stored. In every such directories created, there are four files:

```
$ bcftools isec GiaB.SNPS.vcf.gz HaplotypeCaller.SNPS.vcf.gz -p Giab.HC
$ cat Giab.HC/README.txt 
This file was produced by vcfisec.
The command line was:   bcftools isec  -p Giab.HC GiaB.SNPS.vcf.gz HaplotypeCaller.SNPS.vcf.gz
0000.vcf        for records private to  GiaB.SNPS.vcf.gz
0001.vcf        for records private to  HaplotypeCaller.SNPS.vcf.gz
0002.vcf        for records from GiaB.SNPS.vcf.gz shared by both
0003.vcf        for records from HaplotypeCaller.SNPS.vcf.gz shared by both
```

We can interpret these also as sets and their intersections. This way we are generating pairwise intersections among the etalon GiaB, HaplotypeCaller and Strelka:

```
cd snps/Giab.HC
snps/Giab.HC $ for f in *vcf; do echo -n $f" "; awk '!/^#/{print}' $f|wc -l ; done
0000.vcf 226886       <- missed by HC 
0001.vcf 6213         <- not in GiaB
0002.vcf 2861270      <-
0003.vcf 2861270      <- common SNPs
$ cd ..
$ bcftools isec GiaB.SNPS.vcf.gz Strelka.SNPs.vcf.gz -p Giab.Strelka
$ snps $ cd Giab.Strelka/
$ snps/Giab.Strelka $ for f in *vcf; do echo -n $f" "; awk '!/^#/{print}' $f|wc -l ; done
0000.vcf 47533        <- missed by Strelka
0001.vcf 5569         <- not in GiaB
0002.vcf 3040623      <- 
0003.vcf 3040623      <- common SNPs 
$ cd ..
snps $ bcftools isec HaplotypeCaller.SNPS.vcf.gz Strelka.SNPs.vcf.gz -p HC.Strelka
snps $ cd HC.Strelka/
snps/HC.Strelka $ for f in *vcf; do echo -n $f" "; awk '!/^#/{print}' $f|wc -l ; done
0000.vcf 6674          <- HC only
0001.vcf 185383        <- Strelka only
0002.vcf 2860809       <- 
0003.vcf 2860809       <- common
```

This is not the whole picture to have a 3-sets Venn diagram, in the next step we have to do the same, but this time we are using these intersections. Calculating the complement sets and the common intersectoin region in a similar fashion: 

```
snps $ bgzip -c Giab.Strelka/0003.vcf > GS.common.gz && tabix GS.common.gz
snps $ bcftools isec GS.common.gz HaplotypeCaller.SNPS.vcf.gz -p GS.common.HC 
snps $ bgzip -c Giab.HC/0003.vcf > GH.common.gz && tabix GH.common.gz
snps $ bcftools isec GH.common.gz Strelka.SNPS.vcf.gz -p GH.common.S 
snps $ bgzip -c HC.Strelka/0003.vcf > HS.common.gz && tabix HS.common.gz
snps $ bcftools isec HS.common.gz GiaB.SNPS.vcf.gz -p HS.common.G 
```

Now we have all the files, final step to get the SNP counts for each. As a bonus, we are getting the "common for all" count as a checksum test:

```
snps $ for f in GS.common.HC/*vcf; do echo -n $f" "; awk '!/^#/{print}' $f|wc -l ; done
GS.common.HC/0000.vcf 180954  <- common for GiaB and Strelka
GS.common.HC/0001.vcf 7814    <- only in HaplotypeCaller
GS.common.HC/0002.vcf 2859669 <- common for all
GS.common.HC/0003.vcf 2859669
snps $ for f in GH.common.S/*vcf; do echo -n $f" "; awk '!/^#/{print}' $f|wc -l ; done
GH.common.S/0000.vcf 1601    <- common for GiaB and HaplotypeCaller
GH.common.S/0001.vcf 186523  <- only in Strelka
GH.common.S/0002.vcf 2859669 <- common for all
GH.common.S/0003.vcf 2859669
snps $ for f in HS.common.G/*vcf; do echo -n $f" "; awk '!/^#/{print}' $f|wc -l ; done
HS.common.G/0000.vcf 1140    <- common for HaplotypeCaller and Strelka 
HS.common.G/0001.vcf 228487  <- only in GiaB
HS.common.G/0002.vcf 2859669 <- common for all
HS.common.G/0003.vcf 2859669

```
<br/><img src='/images/GiaB.germline.png'>

What can we learn from this diagram?
- neither Strelka nor HaplotypeCaller caught all the SNPs in the etalon GiaB set
- regarding sensitivity (recall) Strelka fared better in this test, calling 98.46% of SNPs (versus 92.65% for HaplotypeCaller)
- regarding specificity, Strelka and HaplotypeCaller were head-to-head, with 99.82% and 99.78% respectively
- if we are using the intersection of the two callers, the sensitivity (recall) is moderate a 92.60%
- specificity on the other hand is pretty high, 99.96%


#### Indels

By going through all the similar procedure with indels we get the Venn diagram below. Note, there are problems with indels, hence instead of bcftools it is more recommended to use the RTG toolkit for a more serious benchmark test. Still we can see that
 - Strelka sensitivity is 91.93%, specificity is 98.79% (found 91.93% of all indels, and 98.79% of Strelka calls were true positives)
 - ditto HaplotypeCaller sensitivity is 90.83%, specificity is 99.34%
 - intersect sensitivity as expected is the lowest, 89.54%	and specificity - as expected, the highest 99.76%
 - union of calls gives the highest sensitivity (recall) 93.22%	and the lowest (though still pretty high) specificity; 98.40%

<br/><img src='/images/GiaB.germline.indels.png'>

Note, these values are based on high-confidence regions: to get the best calls you still have to filter repeats, low confidence regions (centromeres and telomemres) and highly variable regions like MHC.

