---
title: 'Generating Somatic Mutations For NA12878'
date: 2020-06-08
permalink: /posts/2020/05/NA12878_somatic_generation/
tags:
  - NA12878
  - sarek
  - somatic
  - bamsurgeon
  - singularity
---



```
$ docker login
Login with your Docker ID to push and pull images from Docker Hub. If you don't have a Docker ID, head over to https://hub.docker.com to create one.
Username (szilvajuhos):
Password:
Login Succeeded
$ singularity pull docker://shollizeck/bamsurgeon
WARNING: Authentication token file not found : Only pulls of public images will succeed
INFO:    Starting build...
Getting image source signatures
Copying blob sha256:3386e6af03b043219225367632569465e5ecd47391d1f99a6d265e51bd463a83
 42.08 MiB / 42.08 MiB [====================================================] 0s
[...]

$ singularity shell bamsurgeon_latest.sif
Singularity bamsurgeon_latest.sif:/data2/NA12878/bamsurgeon> addsnv.py
usage: addsnv.py [-h] -v VARFILENAME -f BAMFILENAME -r REFFASTA -o OUTBAMFILE
                 [-s SNVFRAC] [-m MUTFRAC] [-n NUMSNVS] [-c CNVFILE]
                 [-d COVERDIFF] [-z HAPLOSIZE] [-p PROCS]
                 [--picardjar PICARDJAR] [--mindepth MINDEPTH]
                 [--maxdepth MAXDEPTH] [--minmutreads MINMUTREADS]
                 [--avoidreads AVOIDREADS] [--nomut] [--ignoresnps]
                 [--ignoreref] [--force] [--insane] [--single]
                 [--maxopen MAXOPEN] [--requirepaired] [--tagreads]
                 [--skipmerge] [--ignorepileup] [--aligner ALIGNER]
                 [--alignopts ALIGNOPTS] [--tmpdir TMPDIR] [--seed SEED]
addsnv.py: error: argument -v/--varfile is required

```
So far so good.

```
$ find /data1/P2233 -type f -name "Strelka\*somatic\*snvs\*ann\*vcf"| grep CADD | xargs grep HIGH | tee raw.snps| wc -l 
1393
$ awk -F: 'BEGIN{OFS="\t"}$2!~/#/{split($2,cols,"\t");print cols[1],cols[2],cols[2],0.5}' raw.snps| sort -u -V -k1,1 -k2,2n| tee SNPS.tsv |wc -l
671
$ samtools view -@36 -b -o NA12878.chr17.bam ../results/Preprocessing/NA12878N/Recalibrated/NA12878N.recal.bam chr17:
$ samtools index NA12878.chr17.bam
$ mkdir ref &&  cd ref
$ for f in /references/igenomes/Homo\_sapiens/GATK/GRCh38/Sequence/BWAIndex/\* ; do ln -s $f ; done
$ for f in /references/igenomes/Homo\_sapiens/GATK/GRCh38/Sequence/WholeGenomeFasta/\* ; do ln -s $f ; done
$ cd ..
$ singularity shell bamsurgeon\_latest.sif
Singularity bamsurgeon\_latest.sif:/data2/NA12878/bamsurgeon> time addsnv.py -v chr17.snps.tsv -f NA12878.chr17.bam -r ref/Homo\_sapiens\_assembly38.fasta -o NA12878.chr17.somatic.bam -p 12 --force
[...]
real    10m54.014s
user    19m3.553s
sys     6m49.376s

Singularity bamsurgeon\_latest.sif:/data2/NA12878/bamsurgeon> time addsnv.py -v SNPS.tsv -f ../results/Preprocessing/NA12878N/Recalibrated/NA12878N.recal.bam -r ref/Homo\_sapiens\_assembly38.fasta -o NA12878.somatic.bam -p 12 --force
loading donor reads into dictionary...
secondary reads count:0
supplementary reads count:0
loaded 61300 reads, (0 excluded, 0 null or secondary or supplementary--> ignored)

replaced 58798 reads (0 excluded )
kept 0 secondary reads.
kept 0 supplementary reads.
ignored 5399318 non-primary reads in target BAM.

real    390m20.143s
user    659m41.757s
sys     200m1.629s

$ find /data1/P2233 -type f -name "Strelka\*somatic\*indels\*vcf"| grep CADD| xargs grep HIGH | tee raw.indels | wc -l 
356

$ awk -F: -f makeIndels.awk ../raw.indels | sort -u -V -k1,1 -k2,2n| tee indels.tsv |wc -l
253

$ awk -f makelongs.awk /data1/references/annotations/Twist\_Target\_hg38\_updated.liftedover.chr.unique.bed

BEGIN{
  srand();
  OFS="\t"
}
# the input file is a BED with exon coordinates (kinda CCDS bed file from UCSC table browser) 
$3-$2<180 && $3-$2>50 && rand()<0.001 {
  if(rand()>0.5) {  # make an instertion
    chr = $1;
    start = $2;
    end = $2+1;
    cmd = "samtools faidx  /data1/references/igenomes/Homo_sapiens/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38.fasta "$1":"$2"-"$3"|awk '{printf(\"%s \",$1)}'"
    cmd | getline                                                                                                                                                                                                                                 seq = $2$3$4
    print chr,start,end,"0.5\tINS",seq
  } else {
    print $1,$2,$3,"0.5\tDEL"
  }
}

$ time addindel.py -v allindels.tsv -f ../NA12878.somatic.snps.sorted.bam  -r ../ref/Homo_sapiens_assembly38.fasta -o NA12878.somatic.snps.indels.bam -p 12 --force

replaced 17234 reads (0 excluded )
kept 0 secondary reads.
kept 0 supplementary reads.
ignored 0 non-primary reads in target BAM.

real    362m4.551s
user    508m34.553s
sys     88m10.712s

Singularity bamsurgeon_latest.sif:/data2/NA12878/bamsurgeon/indels> samtools sort -@32 -o NA12878.somatic.snps.indels.sorted.bam NA12878.somatic.snps.indels.bam
Singularity bamsurgeon_latest.sif:/data2/NA12878/bamsurgeon/indels> samtools index NA12878.somatic.snps.indels.sorted.bam

We are not doing BAM 2 FASTQ conversion since 
 - RGs are not easy to restore
 - we are missing many reads anyway (Sarek ignores centromeres)
 - it is OK for most of the processes, but for Mutect2

Singularity bamsurgeon_latest.sif:/data2/NA12878/bamsurgeon> time addindel.py -v indels/allindels.tsv -f NA12878.somatic.snps.sorted.bam -r ref/Homo_sapiens_assembly38.fasta -o NA12878.somatic.snps.indels.bam -p 12 --force --aligner mem --alignopts t:42,w:200 --force 2>&1 | tee ins.log

$ samtools view -H NA12878.somatic.snps.indels.sorted.bam | sed 's/NA12878N/NA12878T/g' > T.header.sam
$ samtools reheader -P T.header.sam NA12878.somatic.snps.indels.sorted.bam > NA12878T.bam
$ nextflow run nf-core/sarek -r dev -profile munin --step variantcalling --tools mutect2 --skip_qc all --input simulated_tumor/na12878TNpair.tsv
```
